d1985r0.md

```diff
@@ -414,6 +423,163 @@ reflection.
+template <auto> using takes_auto = void;
+using x = puts_int<takes_auto>; // OK
+using w = puts_auto<takes_int>; // OK??? MSVC, GCC and clang all accept
+```
+
+Both co- and contravariance of function pointers is disallowed completely,
+except for virtual functions, where (pointer and reference) covariant return
+types are allowed, but there are no contravariant parameters.
+
+How about parameter packs?
+
+```cpp
+template <template <typename> typename f> using puts_one = void;
+template <template <typename...> typename f> using puts_var = void;
+template <typename> using takes_one = void;
+template <typename...> using takes_var = void;
+using x = puts_one<takes_var>; // Error! ... behaves covariantly!
+using w = puts_var<takes_one>; // OK, compiles
+```
+
+Turns out parameter packs behave *covariantly* - exactly the opposite of what
+one would expect given the concepts example above.
+
+We have inconsistent behavior across the language. Let's say we did the right
+thing and made `template auto` behave contravariantly, like concepts.
+
+```cpp
+template <template <auto> typename f> using puts_value = void;
+template <template <template auto> typename f> using puts_any = void;
+template <auto> using takes_value = void;
+template <template auto> using takes_any = void;
+using x = puts_value<takes_any>; // OK
+using w = puts_any<takes_value>; // Error because contravariant.
+```
+
+But then, how do we write `apply`? Let's do it for a single argument to avoid
+complications with a covariant `...`:
+
+```cpp
+template <template <template auto T> typename f, template auto arg>
+using apply1 = f<arg>;
+```
+
+The above is correct, but _useless_ - it requires `f`s signature to be
+`template <template auto T>`. What we want to express is that `f` is
+any kind of unary template template, so we can pass in something like
+`template <int x> using int_constant = std::integral_constant<int, x>;`.
+
+Let's call the covariant version of `template auto` (with the meaning "deduce
+this from the argument") `auto template`:
+
+```cpp
+template <template <auto template T> typename f, template auto arg>
+using apply1 = f<arg>;
+```
+
+This is what we want to express (and check at instantiation time), but now
+the `arg`s constraint is spelled differently from `f`'s constraint, and that
+might be very, very difficult to teach.
+
+It also requires us to reserve an additional combination of tokens. Contrast
+what happens if we just made `template auto` behave covariantly (the way
+`auto template` behaves above) if used that place. What do we lose?
+
+At first glance, we lose the ability to define the contravariant meaning of
+`template auto` - a template parameter that *can* bind to anything. But can
+we get that back?
+
+Recall that concepts behave contravariantly. Consider this one:
+
+```cpp
+namespace std {
+  template <template auto arg>
+  concept anything = true;
+};
+```
+
+If we relaxed the rule that only type concepts could appear in the template
```

Comment:
Yes please! That's a bizarre restriction. Do you know the rationale?

Reply:
No, I don't. Separate paper though.